chain-signatures/node/src/rpc.rs

```diff
@@ -0,0 +1,521 @@
+use near_crypto::InMemorySigner;
+use near_fetch::result::ExecutionFinalResult;
+use near_primitives::hash::CryptoHash;
+use serde_json::json;
+use std::collections::VecDeque;
+use std::str::FromStr as _;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::{mpsc, RwLock};
+use url::Url;
+use web3::contract::tokens::Tokenizable as _;
+use web3::ethabi::Token;
+use web3::types::U256;
+
+use k256::elliptic_curve::point::AffineCoordinates;
+use k256::elliptic_curve::sec1::ToEncodedPoint;
+
+pub const PUBLISH_MAX_RETRY: u8 = 6;
+
+enum RpcAction {
+    Publish(crypto_shared::PublicKey, ToPublish),
+}
+
+#[derive(Clone)]
+pub struct RpcChannel {
+    tx: mpsc::Sender<RpcAction>,
+}
+
+impl RpcChannel {
+    pub async fn publish(&self, public_key: crypto_shared::PublicKey, to_publish: ToPublish) {
+        if let Err(err) = self
+            .tx
+            .send(RpcAction::Publish(public_key, to_publish))
+            .await
+        {
+            tracing::error!(%err, "failed to send publish action");
+        }
+    }
+}
+
+pub struct RpcExecutor {
+    near: NearClient,
+    eth: EthClient,
+    action_rx: mpsc::Receiver<RpcAction>,
+    tasks: VecDeque<tokio::task::JoinHandle<()>>,
+}
+
+impl RpcExecutor {
+    pub fn new(near: &NearClient, eth: &EthClient) -> (RpcChannel, Self) {
+        let (tx, rx) = mpsc::channel(1024);
+        (
+            RpcChannel { tx },
+            Self {
+                near: near.clone(),
+                eth: eth.clone(),
+                action_rx: rx,
+                tasks: VecDeque::new(),
+            },
+        )
+    }
+
+    pub async fn run(
+        mut self,
+        contract_state: Arc<RwLock<Option<ProtocolState>>>,
+        config: Arc<RwLock<Config>>,
+    ) {
+        let mut update_interval = tokio::time::interval(Duration::from_millis(3000));
+        let mut action_interval = tokio::time::interval(Duration::from_millis(50));
+        loop {
+            tokio::select! {
+                action = self.action_rx.recv() => {
+                    if let Some(action) = action {
+                        let task = match action {
+                            RpcAction::Publish(public_key, to_publish) => {
+                                execute_publish(self.client(&to_publish.chain), public_key, to_publish)
+                            }
+                        };
+                        self.tasks.push_back(tokio::spawn(task));
+                    } else {
+                        tracing::error!("rpc channel closed unexpectedly");
+                        return;
+                    }
+                }
+                _ = action_interval.tick() => {
+                    // Remove finished tasks
+                    self.tasks.retain(|task| !task.is_finished());
+                }
+                _ = update_interval.tick() => {
+                    self.update_contract(&contract_state).await;
+                    self.update_config(&config).await;
```

Comment:
If you want to keep the shared loop, maybe for easier borrowing, but still avoid blocking signature publishing, you could maybe also spawn this in a separate task. Probably still needs some kind of refactoring for separating out the shared `self`, though.